src/main/scala/com/github/mrpowers/spark/fast/tests/DatasetComparer.scala

```diff
@@ -0,0 +1,115 @@
+import org.apache.spark.sql.{DataFrame, Dataset, Row}
+import org.apache.spark.sql.functions._
+
+import scala.reflect.ClassTag
+
+case class DatasetSchemaMismatch(smth: String) extends Exception(smth)
+case class DatasetContentMismatch(smth: String) extends Exception(smth)
+
+trait DatasetComparer
+    extends TestSuite
+    with DatasetComparerLike { self: Suite =>
+}
+
+trait DatasetComparerLike extends TestSuiteLike {
+
+  def schemaMismatchMessage[T](actualDS: Dataset[T], expectedDS: Dataset[T]): String = {
+    s"""
+Actual Schema:
+${actualDS.schema}
+Expected Schema:
+${expectedDS.schema}
+"""
+  }
+
+  def contentMismatchMessage[T](actualDS: Dataset[T], expectedDS: Dataset[T]): String = {
+    s"""
+Actual DataFrame Content:
+${DataFramePrettyPrint.showString(actualDS.toDF(), 5)}
+Expected DataFrame Content:
+${DataFramePrettyPrint.showString(expectedDS.toDF(), 5)}
+"""
+  }
+
+  def countMismatchMessage(actualCount: Long, expectedCount: Long): String = {
+    s"""
+Actual DataFrame Row Count: '${actualCount}'
+Expected DataFrame Row Count: '${expectedCount}'
+"""
+  }
+
+  def assertSmallDatasetEquality[T](
+    actualDS: Dataset[T],
+    expectedDS: Dataset[T],
+    orderedComparison: Boolean = true
+  ): Unit = {
+    if (!actualDS.schema.equals(expectedDS.schema)) {
+      throw new DatasetSchemaMismatch(schemaMismatchMessage(actualDS, expectedDS))
+    }
+    if (orderedComparison == true) {
+      if (!actualDS.collect().sameElements(expectedDS.collect())) {
+        throw new DatasetContentMismatch(contentMismatchMessage(actualDS, expectedDS))
+      }
+    } else {
+      val actualSortedDF = defaultSortDataset(actualDS)
+      val expectedSortedDF = defaultSortDataset(expectedDS)
+      if (!actualSortedDF.collect().sameElements(expectedSortedDF.collect())) {
+        throw new DatasetContentMismatch(contentMismatchMessage(actualSortedDF, expectedSortedDF))
+      }
+    }
+  }
+
+  def defaultSortDataset[T](ds: Dataset[T]): Dataset[T] = {
+    val colNames = ds.columns.sorted
+    val cols = colNames.map(col(_))
+    ds.sort(cols: _*)
+  }
+
+  def assertLargeDatasetEquality[U](actualDS: Dataset[U], expectedDS: Dataset[U])(implicit UCT: ClassTag[U]): Unit = {
+    if (!actualDS.schema.equals(expectedDS.schema)) {
+      throw new DatasetSchemaMismatch(schemaMismatchMessage(actualDS, expectedDS))
+    }
+    try {
+      actualDS.rdd.cache
+      expectedDS.rdd.cache
+
+      val actualCount = actualDS.rdd.count
+      val expectedCount = expectedDS.rdd.count
+      if (actualCount != expectedCount) {
+        throw new DatasetContentMismatch(countMismatchMessage(actualCount, expectedCount))
+      }
+
+      val expectedIndexValue: RDD[(Long, U)] = zipWithIndex(actualDS.rdd)
+      val resultIndexValue: RDD[(Long, U)] = zipWithIndex(expectedDS.rdd)
+      val unequalRDD = expectedIndexValue
+        .join(resultIndexValue)
+        .filter {
+          case (idx, (o1, o2)) =>
+            !o1.equals(o2)
```

Comment:
I think so.... haha.  I am going to merge this pull request in an open an issue to discuss this issue in more detail!